File: hydra-haskell/src/main/haskell/Hydra/Dsl/Literals.hs
-- | A DSL for constructing Hydra literal values in Haskell

module Hydra.Dsl.Literals where

import Hydra.Core

import Data.Int
import qualified Data.ByteString as B


-- | Create an arbitrary-precision floating point literal
-- Example: bigfloat 3.14159265359
bigfloat :: Double -> Literal
bigfloat = float . FloatValueBigfloat

-- | Create an arbitrary-precision integer literal
-- Example: bigint 9223372036854775808
bigint :: Integer -> Literal
bigint = integer . IntegerValueBigint . fromIntegral

-- | Create a binary data literal
-- Example: binary (B.pack [0x48, 0x65, 0x6C, 0x6C, 0x6F])
binary :: B.ByteString -> Literal
binary = LiteralBinary

-- | Create a boolean literal
-- Example: boolean True
boolean :: Bool -> Literal
boolean = LiteralBoolean

-- | Create a 32-bit floating point literal
-- Example: float32 3.14
float32 :: Float -> Literal
float32 = float . FloatValueFloat32

-- | Create a 64-bit floating point literal
-- Example: float64 3.14159265359
float64 :: Double -> Literal
float64 = float . FloatValueFloat64

-- | Create a floating-point literal with specified precision
-- Example: float (FloatValueFloat32 3.14)
float :: FloatValue -> Literal
float = LiteralFloat

-- | Create a 16-bit signed integer literal
-- Example: int16 32767
int16 :: Int16 -> Literal
int16 = integer . IntegerValueInt16 . fromIntegral

-- | Create a 32-bit signed integer literal
-- Example: int32 42
int32 :: Int -> Literal
int32 = integer . IntegerValueInt32

-- | Create a 64-bit signed integer literal
-- Example: int64 9223372036854775807
int64 :: Int64 -> Literal
int64 = integer . IntegerValueInt64 . fromIntegral

-- | Create an 8-bit signed integer literal
-- Example: int8 127
int8 :: Int8 -> Literal
int8 = integer . IntegerValueInt8 . fromIntegral

-- | Create an integer literal with specified bit width
-- Example: integer (IntegerValueInt32 42)
integer :: IntegerValue -> Literal
integer = LiteralInteger

-- | Create a string literal
-- Example: string "hello world"
string :: String -> Literal
string = LiteralString

-- | Create a 16-bit unsigned integer literal
-- Example: uint16 65535
uint16 :: Int -> Literal
uint16 = integer . IntegerValueUint16 . fromIntegral

-- | Create a 32-bit unsigned integer literal
-- Example: uint32 4294967295
uint32 :: Int64 -> Literal
uint32 = integer . IntegerValueUint32 . fromIntegral

-- | Create a 64-bit unsigned integer literal
-- Example: uint64 18446744073709551615
uint64 :: Integer -> Literal
uint64 = integer . IntegerValueUint64 . fromIntegral

-- | Create an 8-bit unsigned integer literal
-- Example: uint8 255
uint8 :: Int16 -> Literal
uint8 = integer . IntegerValueUint8 . fromIntegral

File: hydra-python/src/main/python/hydra/dsl/literals.py
"""A DSL for constructing literal values."""

from hydra.core import (
    Literal,
    FloatValue,
    IntegerValue,
    LiteralBinary,
    LiteralBoolean,
    LiteralFloat,
    LiteralInteger,
    LiteralString,
    FloatValueBigfloat,
    IntegerValueBigint,
    FloatValueFloat32,
    FloatValueFloat64,
    IntegerValueInt16,
    IntegerValueInt32,
    IntegerValueInt64,
    IntegerValueInt8,
    IntegerValueUint16,
    IntegerValueUint32,
    IntegerValueUint64,
    IntegerValueUint8,
)


def binary(value: bytes) -> Literal:
    """Construct a binary literal."""
    return LiteralBinary(value)


def boolean(value: bool) -> Literal:
    """Construct a boolean literal."""
    return LiteralBoolean(value)


def string(value: str) -> Literal:
    """Construct a string literal."""
    return LiteralString(value)


def float_(value: FloatValue) -> Literal:
    """Construct a float literal."""
    return LiteralFloat(value)


def float32(value: float) -> Literal:
    """Construct a float32 literal."""
    return float_(FloatValueFloat32(value))


def float64(value: float) -> Literal:
    """Construct a float64 literal."""
    return float_(FloatValueFloat64(value))


def bigfloat(value: float) -> Literal:
    """Construct a bigfloat literal."""
    return float_(FloatValueBigfloat(value))


def integer(value: IntegerValue) -> Literal:
    """Construct an integer literal."""
    return LiteralInteger(value)


def int8(value: int) -> Literal:
    """Construct an int8 literal."""
    return integer(IntegerValueInt8(value))


def int16(value: int) -> Literal:
    """Construct an int16 literal."""
    return integer(IntegerValueInt16(value))


def int32(value: int) -> Literal:
    """Construct an int32 literal."""
    return integer(IntegerValueInt32(value))


def int64(value: int) -> Literal:
    """Construct an int64 literal."""
    return integer(IntegerValueInt64(value))


def bigint(value: int) -> Literal:
    """Construct a bigint literal."""
    return integer(IntegerValueBigint(value))


def uint16(value: int) -> Literal:
    """Construct a uint16 literal."""
    return integer(IntegerValueUint16(value))


def uint32(value: int) -> Literal:
    """Construct a uint32 literal."""
    return integer(IntegerValueUint32(value))


def uint8(value: int) -> Literal:
    """Construct a uint8 literal."""
    return integer(IntegerValueUint8(value))


def uint64(value: int) -> Literal:
    """Construct a uint64 literal."""
    return integer(IntegerValueUint64(value))
